packages/midi-input/src/MidiInputPanel.tsx

```diff
@@ -1,87 +1,85 @@
-import { useState } from 'react'
-import { EnumOption, HedronEngine, Input } from '@hedron/engine'
-import {
-  Button,
-  ControlGrid,
-  EnumDropdown,
-  NodeControl,
-  NodeControlInner,
-  NodeControlMain,
-  NodeControlTitle,
-} from '@hedron/ui-core'
-import { midiMessageNames } from '@hedron/midi-manager'
-import { MidiInput, MidiInputOptions } from './MidiInput'
+import { useCallback, useState } from 'react'
+import { HedronEngine, Input } from '@hedron/engine'
+import { Button, ControlGrid, Param, useEngineStore } from '@hedron/ui-core'
+import { MidiInput } from './MidiInput'
 
 interface IProps {
-  input: Input<MidiInputOptions>
+  input: Input
   // TODO: This can be typed as something like HedronEngineWithPlugin<MidiInput>
   engine: HedronEngine
 }
 
-const channelOptions: EnumOption[] = []
-for (let i = 0; i <= 15; i++) {
-  channelOptions.push({ label: `${i + 1}`, value: i })
-}
+const useMidiLearn = (input: Input, engine: HedronEngine) => {
+  const [isLearning, setIsLearning] = useState(false)
 
-const noteOptions: EnumOption[] = []
-for (let i = 0; i < 127; i++) {
-  noteOptions.push({ label: `${i + 1}`, value: i })
+  // TODO: May not need this "as" if we have HedronEngineWithPlugin<MidiInput>
+  const plugin = engine.plugins['midi-input'] as MidiInput
+  const midiManager = plugin.midiManager
+
+  const runMidiLearn = useCallback(async () => {
+    setIsLearning(true)
+    midiManager
+      .midiLearn()
+      .then((event) => {
+        if (!event) {
+          setIsLearning(false)
+          return
+        }
+
+        const store = engine.getStore()
+        const state = store.getState()
+
+        // Update each option node with the learned values
+        input.optionNodeIds.forEach((nodeId) => {
+          const node = state.nodes[nodeId]
+          if (!node) return
+
+          switch (node.key) {
+            case 'channel':
+              state.updateNodeValue(nodeId, event.channel)
+              break
+            case 'note':
+              state.updateNodeValue(nodeId, event.note)
+              break
+            case 'type':
+              state.updateNodeValue(nodeId, event.type)
+              break
+          }
+        })
```

Comment:
this feels a little clunky, would be nice if it was a one liner for each key you want to modify, not really sure what that would look like. `state.updateInputOptionNodeValue(input, 'channel', event.channel)`, or maybe `input.updateOptionValue('channel', event.channel)`

could even have a bulk one that takes an object and updates any keys that have a match, so you could theoretically `input.bulkUpdateOption(event)`